### PR TITLE
:construction: Create prototype of client (EPI-34/EPI-92/EPI-93/EPI-100/EPI-101/EPI-102/EPI-116)

### PR DESCRIPTION
**Pull request by Marius-P1**
This pull request includes several additions and modifications to support new game objects and their corresponding scripts, as well as updates to the build configuration files.

### Additions to Game Objects and Scripts:

* [`assets/objects/json/Ally.json`](diffhunk://#diff-56a9e5b3f7927688de22e810f974650dd17a46b76a010643ac7af6e888a02d9fR1-R34): Added a new JSON definition for the Ally game object.
* [`assets/objects/json/Background.json`](diffhunk://#diff-b79c41d6bbaa22b1afd04f365fde8b2142df8d6ad92df80a31630f3fcd4748a7R1-R38): Added a new JSON definition for the Background game object.
* [`assets/objects/json/Bullet.json`](diffhunk://#diff-ae59337f98963880876f7e5ed8c1f45b0f5cabe844491837736d95710a2702ddR1-R34): Added a new JSON definition for the Bullet game object.
* [`assets/objects/json/Enemy.json`](diffhunk://#diff-72ec05a899abf7ada1750197071328892e810cad9acfb27054ededb7ef4ad388R1-R34): Added a new JSON definition for the Enemy game object.
* [`assets/objects/json/NetworkManager.json`](diffhunk://#diff-2aad92f536c1b630d7b7bb5d89407134deb9becf34aab386e72d962b58e65827R1-R34): Added a new JSON definition for the NetworkManager game object.
* [`assets/objects/json/Player.json`](diffhunk://#diff-2ab04357ab5ebe0ed816fb071679ac954b92e2091597afc61a2ad9e711fd9a75R1-R34): Added a new JSON definition for the Player game object.
* `assets/objects/scripts/Background.cpp` and `assets/objects/scripts/Background.hpp`: Implemented the Background script class to handle background movement. [[1]](diffhunk://#diff-0d7cea661cacde1f3790261f99419dda9b16ce15dee47eebe78f4c0ad043716aR1-R49) [[2]](diffhunk://#diff-a2fe0cbe84710a576efa61a668241ee2fad9991446ac78e18e40b7f506c785c7R1-R37)
* `assets/objects/scripts/NetworkManager.cpp` and `assets/objects/scripts/NetworkManager.hpp`: Implemented the NetworkManager script class to handle network communication and entity synchronization. [[1]](diffhunk://#diff-0f435c294698f0a84db6fa0736e9174af5f12a33e0542f67615f25a46a51c8ecR1-R192) [[2]](diffhunk://#diff-40076f4235c0f826294b6fe32459a80c9fb1f97e90d6fb18f24aea40434fd903R1-R73)

### Build Configuration Updates:

* [`CMakeLists.txt`](diffhunk://#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR10-R20): Included new directories and added required packages (`glm`, `stduuid`, `GSL`, `SFML`).
* [`client/CMakeLists.txt`](diffhunk://#diff-8404a76cc315289a505a0b9e571242cb4f5ba0c122fb86382329a4b714425ac3L3-L11): Updated the client build configuration to link against new libraries and include new source files.
* [`client/dto/CMakeLists.txt`](diffhunk://#diff-ea87aa5cd063991d1e21da1dd69048cd1693d8e92294abf49e7c2c13e80a80a3R1-R31): Added DTO source files to the client build configuration.

### Scene Configuration:

* [`assets/scenes/json/Scene.json`](diffhunk://#diff-7f6b2cffc40967370e70778937c03af33a6a1f6eab194a86a618fff942032fc7R1-R12): Added a new scene configuration including various game objects by their UUIDs.

### DTO Handling:

* [`client/dto/DTODecoder.cpp`](diffhunk://#diff-df07b6a74eee18c5b4246920349be9a46cf648ad4e879365da947b128d61527bR1-R37): Implemented the `DTODecoder` class to decode data transfer objects (DTOs) from binary data.

### CODE REVIEW

**Review by 6im0n — APPROVED** [review]
Good for me !

File: assets/objects/scripts/Background.cpp [hunk 1]
```diff
@@ -0,0 +1,52 @@
+/*
+** EPITECH PROJECT, 2024
+** R-Type-Reborn
+** File description:
+** Background.cpp
+*/
+
+#include "Background.hpp"
+
+Background::Background(IObject *owner, const json::IJsonObject *data) : CPPMonoBehaviour(owner) {}
```
Comment (axel-eck) [inline assets/objects/scripts/Background.cpp]:
```suggestion
Background::Background(IObject *owner, const json::JsonObject *data) : CPPMonoBehaviour(owner, data) {}
```

File: assets/objects/scripts/Background.hpp [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+#define BACKGROUND_HPP
+
+#include <chrono>
+#include "common/components/CPPMonoBehaviour.hpp"
+#include "common/components/Transform.hpp"
+#include "common/json/JsonObject.hpp"
+
+class Background final : public CPPMonoBehaviour {
+    public:
+        Background(IObject *owner, const json::IJsonObject *data);
```
Comment (axel-eck) [inline assets/objects/scripts/Background.hpp]:
```suggestion
        Background(IObject *owner, const json::JsonObject *data);
```

File: client/dto/entity/AEntityDTO.hpp~HEAD
Comment (axel-eck) [inline client/dto/entity/AEntityDTO.hpp~HEAD:1]: ?

File: assets/scenes/json/Scene.json
Comment (axel-eck) [inline assets/scenes/json/Scene.json:1]: Weird indent

File: client/dto/DTORegistry.cpp [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+#include "entity/EntityDeletionDTO.hpp"
+#include "entity/EntityPositionDTO.hpp"
+#include "player/PlayerActionStartDTO.hpp"
+#include "player/PlayerActionStopDTO.hpp"
+
+#include <typeinfo>
+
+DTORegistry::DTORegistry()
+{
+    this->_dtos = {
```
Comment (axel-eck) [inline client/dto/DTORegistry.cpp:19]: can't this be static ?

File: client/dto/entity/AEntityDTO.hpp~epi-34-create-prototype-of-server
Comment (axel-eck) [inline client/dto/entity/AEntityDTO.hpp~epi-34-create-prototype-of-server:1]: ?

File: client/dto/player/APlayerDTO.hpp~HEAD
Comment (axel-eck) [inline client/dto/player/APlayerDTO.hpp~HEAD:1]: ?

File: client/dto/player/APlayerDTO.hpp~epi-34-create-prototype-of-server
Comment (axel-eck) [inline client/dto/player/APlayerDTO.hpp~epi-34-create-prototype-of-server:1]: ?

File: r-type_client
Comment (axel-eck) [inline r-type_client:1]: Should not push compiled binaries

**Review by axel-eck — CHANGES_REQUESTED** [review]

File: assets/objects/scripts/Background.cpp [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+Background::Background(IObject *owner, const json::JsonObject *data) :
+    CPPMonoBehaviour(owner, data) {}
+
+void Background::start() {
+    startTime = std::chrono::high_resolution_clock::now();
+    actualTime = std::chrono::high_resolution_clock::now();
+    auto *transform = getParentComponent<Transform>();
+    transform->setPosition(Vector3(0, 0, 0));
+}
+#include <iostream>
```
Comment (axel-eck) [inline assets/objects/scripts/Background.cpp]: 👀

**Review by axel-eck — APPROVED** [review]

**Review by 6im0n — APPROVED** [review]

**Pull request merged**